### PR TITLE
Add Cooperative Storage

### PR DESCRIPTION
**Pull request by hayesgm**
This patch adds back cooperative storage, which simply isolates each `read` and `write` call in `QuarkScript` to pulling the active nonce and isolating storage keys on that value.

Notes:
  * This is cooperative. A malicious script is free to change any stored data.
  * After a nested quark operation, the nonce value is cleared and thus will not be properly read. We raise an error in this case.

### CODE REVIEW

File: src/quark-core/src/QuarkScript.sol [hunk 1]
```diff
@@ -114,11 +116,22 @@ abstract contract QuarkScript {
...
+    // Returns a key isolated to the active nonce of a script
+    // This provide cooperative isolation of storage between scripts.
+    function getNonceIsolatedKey(bytes32 key) internal view returns (bytes32) {
+        bytes32 nonce = getActiveNonce();
+        if (nonce == bytes32(0)) {
+            revert NoActiveNonce();
+        }
+        return keccak256(abi.encodePacked(nonce, key));
+    }
+
```
Comment (fluffywaffles) [inline src/quark-core/src/QuarkScript.sol:134]: nice!!

**Review by fluffywaffles — APPROVED** [review]

**Comment by kevincheng96** [comment]
> * After a nested quark operation, the nonce value is cleared and thus will not be properly read. We raise an error in this case.

An alternative approach to address this would be to, in the wallet, cache the current `activeNonce` and set it back to this value after the callcode. Not saying we should do this, but in a `tstore` world this is basically free and would remove this weird edge case.

**Review by kevincheng96 — APPROVED** [review]
LGTM

**Comment by hayesgm** [comment]
> > * After a nested quark operation, the nonce value is cleared and thus will not be properly read. We raise an error in this case.
> 
> An alternative approach to address this would be to, in the wallet, cache the current `activeNonce` and set it back to this value after the callcode. Not saying we should do this, but in a `tstore` world this is basically free and would remove this weird edge case. It's basically what we used to do in the `QuarkStateManager`.

Yeah, though scripts could also cache it themselves. Mostly here I want the scripts to be written pretty seamlessly, but they are free for more complex behaviors if they want to handle nested calls.

**Pull request merged**